src/source-updater.js

```diff
@@ -24,54 +24,144 @@ const actions = {
   },
   callback: (callback) => (type, sourceUpdater) => {
     callback();
+  },
+  duration: (duration) => (sourceUpdater) => {
+    sourceUpdater.mediaSource.duration = duration;
   }
 };
 
 const updating = (type, sourceUpdater) => {
   const sourceBuffer = sourceUpdater[`${type}Buffer`];
 
-  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queue[type].pending;
+  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queuePending[type];
 };
 
-const shiftQueue = (type, sourceUpdater) => {
-  const queue = sourceUpdater.queue[type];
+const processMediaSourceQueueEntry = (index, sourceUpdater) => {
+  const queueEntry = sourceUpdater.queue[index];
+  const types = queueEntry.type;
+  const bufferInUse =
+    types.reduce((acc, type) => acc || updating(type, sourceUpdater), false);
 
-  if (updating(type, sourceUpdater) || !queue.actions.length || !sourceUpdater.started_) {
-    return;
+  if (bufferInUse) {
+    // block until all buffers are ready for use
+    return false;
   }
 
-  const action = queue.actions.shift();
+  // remove the entry since it'll be processed
+  sourceUpdater.queue.splice(index, 1);
+  // when the entry needs multiple source buffer types, an individual type doesn't need to
+  // be passed for the action
+  queueEntry.action(sourceUpdater);
```

Comment:
it looks like this method and the shiftQueue condition on line 92 are doing the same things -- but since we splice out the action, shift queue is processing an action that has already been processed?

Reply:
The shiftQueue allows the next queue entries to process after this action is finished.

Reply:
I like `allows the next queue entries to process after this action is finished.` as a comment for line 92